### ISSUE TITLE
Add BIP44 Support

### ISSUE THREAD

**Issue by christroutner** [Bitcoin-com/bch-cli-wallet#6]
This Issue is tagged with a bounty of **2 BCH**. Once a Pull Request (PR) has been verified to meet the acceptance criteria, and merged into the master branch, the funds will be released to the developer.

## Overview
[BIP44](https://github.com/bitcoin/bips/blob/master/bip-0044.mediawiki#Address_gap_limit) is an HD wallet standard that describes how a HD wallet should generate new addresses to maximize privacy. The primary way in which the current implementation of bch-cli-wallet differs from the standard is in regard to the [address gap limit](https://github.com/bitcoin/bips/blob/master/bip-0044.mediawiki#address-gap-limit).

## Specifications
The scope of this Issue is to refactor the [update-balances](https://github.com/Bitcoin-com/bch-cli-wallet/blob/master/src/commands/update-balances.js) command to bring it into line with BIP44. This includes the following features:

- The `generateAddress()` method currently generates a single address. It should instead generate 20 addresses and return them as an array.
- The `getAddressData()` method retrieves details on a single address at a time.  This should be expanded to use [the new BITBOX bulk call](https://developer.bitcoin.com/bitbox/docs/address#details) to retrieve data on 20 addresses at a time.
- The current implementation stops at `walletInfo.nextAddress` number of addresses. New functionality should continue to query address data in 20-address batches until a block of 20 addresses returns no transaction history.

Before starting work on this Issue, developers should add a comment to this issue with a link to your fork of the code base, to broadcast that you are starting work and so that others can follow the progress of your work. If work pauses for more than 48 hours, other developers are encouraged to pick up this Issue.

## Acceptance Criteria
- Unit tests should be updated, with any needed new test cases added.
- Test coverage percentage should remain the same or increase. The percentage can be seen by running `npm test`.

## Guidance
Guidance can be obtained by leaving comments below, or reaching out to @christroutner via the [Discord developer chat room](https://discord.gg/vzKSTeP).

**Comment by lightswarm124** [comment]
https://github.com/BitcoinBay/bch-cli-wallet

**Comment by lightswarm124** [comment]
> https://github.com/BitcoinBay/bch-cli-wallet

generateAddress() and getAddressData() seem to be working. the json wallet file can now updates the nextAddress value to the next available unused address in the 20 address request limit

will work on looping this process for step 3

**Comment by christroutner** [comment]
Great work!

**Comment by christroutner** [comment]
Just reviewed your latest updates in `master`. Here are my notes:

- Looks like the BITBOX mock for Address.details() returns an array, though not sure if the mocked data still reflects real data returned. Let me know if you need any help with mocking for unit tests.

-The `nextAddress` value in the wallet json file does not appear to be getting updated accurately. Addresses with balances in the wallet JSON file will be greater than the `nextAddress` value.

Other than that, great work. The wallet is handling larger quantities of addresses without running into rate-limits of rest.bitcoin.com.

Again, let me know if you need any assistance with updating the unit tests to reflect the new code changes.

**Comment by christroutner** [comment]
This bounty has been fulfilled.
